website/docs/tutorial/02-making-improvements.md

```diff
@@ -0,0 +1,501 @@
+2. **No crash risk**: `guess` will never encounter a missing number
+3. **Better user experience**: Players can start guessing immediately
+
+Let's also simplify our `reset` function to use the new helper:
+
+```rust
+/// Update the number. Only callable by admin.
+pub fn reset(env: &Env) {
+    Self::require_admin(env);
+    Self::set_random_number(env);
+}
+```
+
+Much cleaner! The logic is now centralized in our helper function. Note that this is still a public function, see the `pub`? The distinction between "public" and "private" might seem confusing here. Let's run the application and it should clear everything up:
+
+```bash
+$ npm start
+```
+
+Click over to `&lt;/&gt; Debugger` if you're not there already and select the `guess_the_number` contract. You'll see that `reset` is listed here, but `set_random_number` is not.
+
+Our `reset` method is available to be called by code _outside_ our contract because we opted in to it being a public method with the `pub` keyword. Our `set_random_number` is private by default, it's not visible to the outside world. It's not listed in the Contract Explorer. It's not listed in the CLI help either:
+
+```bash
+$ stellar contract invoke --id guess_the_number --source me --network local -- help
+Commands:
+  reset    Update the number. Only callable by admin.
+  guess    Guess a number between 1 and 10
+  upgrade  Upgrade the contract to new wasm. Only callable by admin.
+  help     Print this message or the help of the given subcommand(s)
+```
+
+It would error if you tried to invoke it:
+
+```bash
+$ stellar contract invoke --id guess_the_number --source me --network local -- set_random_number
+error: unrecognized subcommand 'set_random_number'
+```
+
+#### Wait, So Anyone Can Call Reset?
+
+Nope! Just because we made it public, we still require authentication so only admins can call it. Rust's idea of public vs private handles "where" the functions can be called. You still need to handle "who" calls it. That's why we set the contract admin in it's constructor method and check it with `Self::require_admin(env);`.
+
+You can try this out by invoking it from the Contract Explorer in your browser. The admin is `me`, but you didn't import that account into your browser wallet. Go ahead and hit `Submit` on the `reset` function.
+
+You could also try this out in the CLI. Create a non-admin identity to see how it fails:
+
+```bash
+$ stellar keys generate bob --network local --fund
+✅ Key saved with alias bob in ".config/stellar/identity/bob.toml"
+✅ Account bob funded on "Standalone Network ; February 2017"
+
+$ stellar contract invoke --id guess_the_number --source bob --network local -- reset
+❌ error: Missing signing key for account GDAQWVA6REGN47BBCFY6SGQ4YTIGMDZZFHDOVUZXMVRAAT6OEZGCACGH
+```
+
+The account called `me` is the admin, Bob is just a regular user. `me` can call `reset`, Bob gets an error.
+
+### Bonus Step 3: Go Pro and Save Bytes with `unsafe`
```

Comment:
I am really not sure about showing that. Why not just using `unwrap`? `unwrap_unchecked` goes a level deeper but at the cost of using `unsafe`. I think people should discover super patterns like that on their own otherwise we should put some big read caveats about not abusing that.